rules/coding-style/src/Rector/String_/UseClassKeywordForClassNameResolutionRector.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\CodingStyle\Rector\String_;
+
+use Nette\Utils\Strings;
+use PhpParser\Node;
+use PhpParser\Node\Expr\BinaryOp\Concat;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Scalar\String_;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+
+/**
+ * @see \Rector\CodingStyle\Tests\Rector\String_\UseClassKeywordForClassNameResolutionRector\UseClassKeywordForClassNameResolutionTest
+ */
+final class UseClassKeywordForClassNameResolutionRector extends AbstractRector
+{
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Use `class` keyword for class name resolution in string instead of hardcoded string reference', [
+            new CodeSample(
+                <<<'PHP'
+$this->write(
+                '<?php if (App\SomeClass::createCache($netteCacheStorage, %var,  $this->global->cacheStack, %node.array?)) { ?>',
+                md5('md5')
+            );
+PHP
+                ,
+                <<<'PHP'
+$this->write(
+                '<?php if (' . \App\SomeClass::class . '::createCache($netteCacheStorage, %var,  $this->global->cacheStack, %node.array?)) { ?>',
+                \md5('md5')
+            );
+PHP
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [String_::class];
+    }
+
+    /**
+     * @param String_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (substr_count($node->value, '::') !== 1) {
+            return null;
+        }
+        // a possible static call reference
+        [$before, $possibleClass, $after] = Strings::split($node->value, "/([\\\\a-zA-Z0-9_\x80-\xff]*)::/");
+
+        if (! $possibleClass || ! class_exists($possibleClass)) {
+            return null;
+        }
+
+        $classConstFetch = new ClassConstFetch(new FullyQualified(ltrim($possibleClass, '\\')), 'class');
+
+        $concat = new Concat($classConstFetch, new String_('::' . $after));
+        if (! empty($before)) {
+            $concat = new Concat(new String_($before), $classConstFetch);
+            $concat = new Concat($concat, new String_('::' . $after));
+        }
```

Comment:
@TomasVotruba, I wanted to use something like this
```
        $concat = new Concat($classConstFetch, new String_('::' . $after));
        if (! empty($before)) {
            $concat = new Concat(new String_($before), $concat);
        }
```
but for some reason it generates unexpected (to me :)) string
`'<?php if (' . (\Rector\CodingStyle\Tests\Rector\String_\UseClassKeywordForClassNameResolutionRector\Fixture\SomeClass::class . '::createCache($netteCacheStorage, %var,  $this->global->cacheStack, %node.array?)) { ?>')`

Reply:
Could you write it in a diff what you want and what's changed?

```diff
+$value . 'String'

Reply:
The comment was updated with diffs

Reply:
I still can't see the problem (too much nette irelevant code).
If it's extra `(` . `)` than it's irelevant, as concat basically joins it all.

Reply:
Maximum simplification:
```diff
-'<?php if (' . (SomeClass::class . '::method()) { ?>')
+'<?php if (' . SomeClass::class . '::method()) { ?>'
```
Doesn't look like it just joins...

Reply:
Does reordering concat works? I think they're working the other way around.

Reply:
If I concat in "proper" order everything works well.

Reply:
Basically... this code works well but doesn't look that optimized:
```
        $concat = new Concat($classConstFetch, new String_('::' . $after));
        if (! empty($before)) {
            $concat = new Concat(new String_($before), $classConstFetch);
            $concat = new Concat($concat, new String_('::' . $after));
        }
```
This code looks more optimized and from the first view should work fine but it doesn't
```
        $concat = new Concat($classConstFetch, new String_('::' . $after));
        if (! empty($before)) {
            $concat = new Concat(new String_($before), $concat);
        }
```
I'm just wondering why it doesn't and I don't mind to leave it as it is now :)